Fix #3245,#2193,#2941,#5143  - Timeline: Migrate to vis.js

jasonex7 opened the pull request:
Fixes #3245 

- Using vis.js 5.1. It was recommended to upgrade to v6 which it should be out in the next few days. It was told that it will be 95% compatible to current 5.1 release so I have only to replace 0-timeline.min.js. (https://github.com/visjs/vis-timeline)
- TimelineEvent, TimelineGroup, TimelineModel, TimelineModificationEvent, TimelineSelectEvent now use generics.
- TimelineEvent has now 3 more properties to fine tunning editable: editableTime, editableGroup and editableRemove.
- TimelineGroup has now styleClass property similar to the one in TimelineEvent.
- Properties deprecated in favor of others: axisOnTop, groupsChangeable, groupsWidth, groupMinHeight, snapEvents, timeChangeable.
- Properties deprecated with NO replacement: dragAreaWith (use css classes instead), unselectable, groupsOnRight, animate, animateZoom, showButtonNew, showNavigation.
- New properties: maxHeight, orientationAxis, orientationItem, editableAdd, editableTime, editableGroup, editableRemove, editableOverrideItems, eventHorizontalMargin, eventVerticalMargin, groupStyle, snap, clickToUse.
- New extender property to specify a javascript function to extend configuration of vis.js timeline component similar to extender property in <p:chart> component.

jasonex7 committed:
Closes #3245 - Timeline: Migrate to vis.js
- Using vis.js 5.1. It was recommended to upgrade to v6 which it should be out in the next few days. It was told that it will be 95% compatible to current 5.1 release so I have only to replace 0-timeline.min.js.
- TimelineEvent, TimelineGroup, TimelineModel, TimelineModificationEvent, TimelineSelectEvent now use generics.
- TimelineEvent has now 3 more properties to fine tunning editable: editableTime, editableGroup and editableRemove.
- TimelineGroup has now styleClass property similar to the one in TimelineEvent.
- Properties deprecated in favor of others: axisOnTop, groupsChangeable, groupsWidth, groupMinHeight, snapEvents, timeChangeable.
- Properties deprecated with NO replacement: dragAreaWith (use css classes instead), unselectable, groupsOnRight, animate, animateZoom, showButtonNew, showNavigation.
- New properties: maxHeight, orientationAxis, orientationItem, editableAdd, editableTime, editableGroup, editableRemove, editableOverrideItems, eventHorizontalMargin, eventVerticalMargin, groupStyle, snap, clickToUse.
- New extender property to specify a javascript function to extend configuration of vis.js timeline component similar to extender property in <p:chart> component.

jasonex7 commented:
I recommend create a new facet called maybe "navigator" or "menu" so we can add custom buttons to add functionality that was deprecated by properties `showButtonNew` and `showNavigation` similiar to this [example](https://visjs.github.io/vis-timeline/examples/timeline/interaction/navigationMenu.html). What do you think @melloware?

christophs78 commented:
Happy to see this PR. Just started yesterday on moving timeline to java.time.LocalDateTime and noted the old library. I'll have to redo some work after this PR is merged, but I think it is worth the work.

melloware commented on src/main/java/org/primefaces/component/timeline/TimelineBase.java:
```diff
@@ -172,21 +182,29 @@ public void setBrowserTimeZone(Object browserTimeZone) {
 
     public String getHeight() {
-        return (String) getStateHelper().eval(PropertyKeys.height, "auto");
+        return (String) getStateHelper().eval(PropertyKeys.height, null);
```
Is "auto" no longer a valid option in the new timeline?  What are the options for height?

jasonex7 replied:
No, "auto" is no longer a valid options, that was something that give me an error on tests. If the property is omitted it behaves like auto (see next image of description of height property). 

![image](https://user-images.githubusercontent.com/16229253/64270195-ca26c980-cf00-11e9-9d17-232c2e54e3a4.png)

jasonex7 committed:
Closes #3245 - Timeline: Migrate to vis.js
- Restore functionality of move function that takes a factor (positive right, negative left) as first parameter to move timeline. The new optionals parameters (options and callback) stands.
- Fixes documentation of zoom function and TimelineEvent editableRemove property.

melloware commented on src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java:
```diff
@@ -119,44 +121,72 @@ public void beforePhase(PhaseEvent event) {
+                        }
+                    }
+                    if (foundGroup != null) {
+                        //If groups was not setted in model then order by content.
```
"setted" should be "set"

jasonex7 replied:
Sorry my bad English on that 😅

melloware commented on src/main/java/org/primefaces/component/timeline/TimelineBase.java:
```diff
@@ -50,13 +54,18 @@
+        editableGroup,
+        editableTime,
+        editableOverrideItems,
         selectable,
```
Need to update /docs/7.1/components/timeline to add and remove all these properties.

jasonex7 replied:
By now I don't remove from taglib to mantain compatibility with existing xhtml code, so I only removed these from `TimelineBase`. I documented on taglib that the properties was deprecated.

melloware replied:
That is good but the actual Markdown doc need to be update in the /docs/7.1/ folder.

jasonex7 replied:
> That is good but the actual Markdown doc need to be update in the /docs/7.1/ folder.

That's a new world for me haha. Check this new commit.

melloware commented on src/main/java/org/primefaces/component/timeline/TimelineRenderer.java:
```diff
@@ -85,42 +86,97 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
+        if (groupsSize > 0) {
+            writer.write(",groups:[");
+            for (int i = 0; i < groupsSize; i++) {
+                //If groups was not setted in model then order by content.
```
"setted" = "set"

melloware commented on src/main/java/org/primefaces/component/timeline/TimelineRenderer.java:
```diff
@@ -141,174 +197,163 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
-            fsw.write(",\"editable\":null");
+        if (event.isEditableTime() != null || event.isEditableGroup() != null || event.isEditableRemove() != null) {
+            fsw.write(", editable: {");
+            boolean wasSetted = false;
```
wasSetted = isSet

melloware commented on src/main/java/org/primefaces/component/timeline/TimelineRenderer.java:
```diff
@@ -141,174 +197,163 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
+        if (event.getGroup() != null) {
+            for (TimelineGroup<?> group : groups) {
                 if (group.getId() != null && group.getId().equals(event.getGroup())) {
                     foundGroup = group;
```
if (Objects.equals(group.getId(), event.getGroup)) {

melloware commented:
> I recommend create a new facet called maybe "navigator" or "menu" so we can add custom buttons to add functionality that was deprecated by properties `showButtonNew` and `showNavigation` similiar to this [example](https://visjs.github.io/vis-timeline/examples/timeline/interaction/navigationMenu.html). What do you think @melloware?

I like that idea.

jasonex7 committed:
Closes #3245 - Timeline: Migrate to vis.js
- Readability.

melloware commented:
Adding link to this ticket which we should also do with this upgrade: https://github.com/primefaces/primefaces/issues/2941

jasonex7 committed:
Closes #3245 - Timeline: Migrate to vis.js
- Documentation of added/deprecated/removed properties.
- Documentation of an example using extender property.

melloware commented:
I just assigned https://github.com/primefaces/primefaces/issues/2193 to make the TimeLine RTL aware! Should be easy addition for you. 😄

jasonex7 commented:
> I just assigned #2193 to make the TimeLine RTL aware! Should be easy addition for you. 😄

You almost read my mind haha. I'm finishing menu facet supporting a new dir (ltr or rtl) property.

jasonex7 committed:
Closes #3245 - Timeline: Migrate to vis.js
- Closes #2193 - TimeLine add RTL support.
- New menu facet to add a custom menu.
- New loading facet to add a custom loading screen using loadingTemplate of vis.js.
- Remove use of timeChangeable and editable properties in 1-timeline.js.
- Documentation of dir property and examples of menu and loading facets.

jasonex7 commented:
> Adding link to this ticket which we should also do with this upgrade: #2941

In vis.js documentation says that localization depends on the localization of `moment.js` library. So the user must load a localized version of moment.js (like [this example](https://visjs.github.io/vis-timeline/examples/timeline/other/localizationCustom.html)) and use `locale` property on `<p:timeline>` or set that locale on the jsf view.

> ### Localization
> Timeline can be localized. For localization, Timeline depends largely on the localization of moment.js. Locales are not included in vis.js by default. To enable localization, moment.js must be loaded with locales. Moment.js offers a bundle named "moment-with-locales.min.js" for this and there are various alternative ways to load locales.

melloware commented:
> In vis.js documentation says that localization depends on the localization of `moment.js` library. So the user must load a localized version of moment.js (like [this example](https://visjs.github.io/vis-timeline/examples/timeline/other/localizationCustom.html)) and use `locale` property on `<p:timeline>` or set that locale on the jsf view.

Maybe we should put this information on the /docs/7.1/ page for users who will definitely ask how to localize it.

jasonex7 commented:
> Maybe we should put this information on the /docs/7.1/ page for users who will definitely ask how to localize it.

OK, I will work on putting that on docs. I also open the issue https://github.com/visjs/vis-timeline/issues/78 on vis.js to ask for localization of delete button, so maybe in a next release (hope in v6 😄) we should update that on docs.

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js #5113
- Closes #2941 - TimeLine: Add localization to the wiki.

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js #5113
- Documentation of Client Side API.

melloware commented:
@jasonex7 this was just fixed this morning: https://github.com/primefaces/primefaces/issues/5118

Can you review and make the same fix in your branch?

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js #5113
- Fix: The id of the new added item was not calculated correctly.
- Fix: The listeners of add, changed, edit and delete events wasn't updating UI after call ajax behavior methods.
- Documentation: Fix example of extender use case.

jasonex7 commented:
> @jasonex7 this was just fixed this morning: #5118
> 
> Can you review and make the same fix in your branch?

I think that is not an issue, but a problem of lack of documentation. The documentation of timeline component when it was on PrimeFaces Extensions was more explained than when it was moved to PrimeFaces. See my comment https://github.com/primefaces/primefaces/issues/5118#issuecomment-528891817

melloware commented:
I see that. Thanks for investigating.

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js #5113
- Fix error on edit callback. Missing this keyword 🤦‍♂ .

jasonex7 committed:
Merge commit 'e8f2ba2931e40b3d7eae2f3fd8ad77073ffc30a4' into visjs-timeline

Conflicts resolved on:
	src/main/java/org/primefaces/component/timeline/TimelineBase.java
	src/main/resources/META-INF/resources/primefaces/timeline/1-timeline.js

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js #5113
- Fix css of timeline menu facet to align to consider vis-tooltip.

jasonex7 committed:
Fix #3245,#2193,#2941 - Timeline: Migrate to vis.js
- Added capability to show tooltips by new property title in TimelineEvent and TimelineGroup and new properties in <p:timeline>.
- New eventTitle facet to provide custom content for event's titles.
- Documentation for this new capability.

jasonex7 committed:
Fix #3245,#2193,#2941,#5143 - Timeline: Migrate to vis.js
- Fixes #5143 - Timeline: Support for multiple events with same data

melloware reviewed (approved):
This is a huge improvement and will prevent future bugs by introducing the id value. Great job!

jasonex7 committed:
Merge commit '9228ddd2dad4335fcce5ed5af00fd60ecb78df4f' into visjs-timeline
- Merge commit that fixes #5149

Conflicts resolved in:
	src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java

jasonex7 commented:
I've just merged PR #5157 that fixes #5149 in this PR.

jasonex7 committed:
Merge commit '3d985f424709e22018ccac70411197e85acaa8fc' into visjs-timeline

tandraschko commented:
V6 already released?
Does the migration break existing apps?

melloware commented:
@tandraschko we are still awaiting the release of v6 which fixes an IE11 issue.  @jasonex7 has done eveyrthing he can to make it backwards compatible and marked which methods are deprecated in both the taglib and code.  Similar to how we did Schedule upgrade in 7.0 I think we will just have to put these notes in the Migration Guide for which props are deprecated or replaced etc.

tandraschko commented:
any news?

melloware commented:
I am monitoring their releases and they have not released 6.0 yet.  They have fixed a lot of issues and @jasonex7 has submitted quite a few PR's that have been accepted.  It should be soon though.

jasonex7 commented:
As a comment, we're using the current vis-timeline build in a production app and we have no problems and works perfectly in IE11 thanks to visjs/vis-timeline#72 got fixed. So maybe, a custom build could be included with this and others PR that have been submitted to visjs/vis-timeline repo, what do you think?

melloware commented:
@jasonex7 I am OK with that so it will let us migrate this PR and allow you/us to work on the showcase examples migration from PF Extensions.  then before we cut 7.1 or whenever its released we can just push the final version of vis timeline.  So I approve.

tandraschko commented:
thats ok for me now.
we should then upgrade the scripts ASAP when its released.

Can you update the PR with a snapshot build from vis?

jasonex7 committed:
Merge commit '0321f25cf7d26839f72474248f8c67b75832527f' into visjs-timeline

jasonex7 committed:
Fix #3245,#2193,#2941,#5143 - Timeline: Migrate to vis.js
- Update vis-timeline to v6.0.0-SNAPSHOT from commit https://github.com/visjs/vis-timeline/tree/9315d243e5a768759d9c7a20162b3901a09745c3.

jasonex7 commented:
> Can you update the PR with a snapshot build from vis?

Done.

melloware commented:
Thanks @jasonex7 its now merged!

tandraschko commented:
hey guys (@jasonex7 @melloware)
if you open the showcase (http://localhost:8080/showcase/ui/data/timeline/linked.xhtml)
and zoom or drag something, a endless loop of AJAX requests will be fired
can you please fix that?

tandraschko commented:
And please also change the migrationguide

melloware commented:
Migration Guide has been updated.

jasonex7 commented:
> if you open the showcase (http://localhost:8080/showcase/ui/data/timeline/linked.xhtml)
> and zoom or drag something, a endless loop of AJAX requests will be fired
> can you please fix that?

I've investigated and that error appears also with the previous timeline version (chap-links-library). There are plenty of ajax requests, but finally they eventually stopped. I will create another issue to work on that.
